HGET: Version 1.3 w/ TCP-IP 1.1 support and HTTPS support

ducasp opened the pull request:
- Check if TLS is supported by UNAPI implementation
- Check if UNAPI implementation supports safe TLS (validating host certificate)
- If protocol is HTTPS and UNAPI implementation supports TLS, open connection using TLS
- Added /u to allow UNSAFE https transfers if adapter supports safe TLS connections
- Added /n to allow to not check hostname if adapter supports safe TLS connections
- There is no need for waiting for an extra tick after calling TCPIP_WAIT. Implementations will make sure of that, if they need it. 
- Changed the progress indication, now a progress bar that is increased in 4% increments. This is easier on CPU/VDP waits and allow faster adapters to get more data instead of waiting for screen to be updated. If file size is unknown, a star-like movement is show to indicate program is not stuck.

Konamiman commented on SRC/NETWORK/hget.c:
```diff
@@ -1409,9 +1487,49 @@ void CreateLocalFile()
+				  pop  HL
+				  pop  AF
+				__endasm;
+			}
```
Please move this chunk of assembler to its own function with a descriptive name, this way the code will be easier to read.

Konamiman commented on SRC/NETWORK/hget.c:
```diff
@@ -1500,14 +1618,29 @@ void ResolveServerName()
+	if (bHTTPS)
+	{
+		if (bCheckCertificate)
+			TcpConnectionParameters->flags = TcpConnectionParameters->flags | 8 ; //set flags properly for TLS certificate validation
```
Please convert that 8 to a constant, e.g. `TCPFLAGS_VERIFY_CERTIFICATE`. This way the code reads better and you don't need the comment.

ducasp replied:
Done :)

Konamiman commented on SRC/NETWORK/hget.c:
```diff
@@ -1500,14 +1618,29 @@ void ResolveServerName()
+	else
+	{
+		TcpConnectionParameters->flags = 0 ; //set flags properly for regular tcp
+		TcpConnectionParameters->hostName =  0; //clear host name buffer
```
I think this comment, and the other ones above in `->hostName =`, are redundant. It's clear enough what the code does without them.

ducasp replied:
Done :)

Konamiman commented on SRC/NETWORK/hget.c:
```diff
@@ -1526,15 +1659,83 @@ void OpenTcpConnection()
         Terminate(strNoNetwork);
     } else if(regs.Bytes.A != 0) {
         sprintf(Buffer, "Unexpected error when opening TCP connection (%i)", regs.Bytes.A);
+		if ((bHTTPS)&&(regs.Bytes.A == ERR_NO_CONN)) //B should have the reason
```
`B should have the reason` - it's C, right?

ducasp replied:
Done :)

Konamiman commented on SRC/NETWORK/hget.c:
```diff
@@ -528,18 +562,37 @@ int CharIsWhitespace(char ch)
+        strcpy(domainName, url + 8);
+		bHTTPS = true;
+		TcpConnectionParameters->remotePort = HTTPS_DEFAULT_PORT;
+		TcpConnectionParameters->flags = TcpConnectionParameters->flags | 4 ; //set flags properly for TLS
```
Please convert that 8 to a constant, e.g. `TCPFLAGS_USE_TLS`. This way the code reads better and you don't need the comment.

ducasp replied:
Done :)

Konamiman commented on SRC/NETWORK/hget.c:
```diff
@@ -174,7 +186,8 @@ byte authenticationSent;
 int emptyLineReaded;
-long contentLength;
+long contentLength,blockSize,currentBlock;
+bool bFirstUpdate;
```
Ok, I understand that hungarian notation is commonly used in the world of C, but since I haven't used it for this program I'd prefer you don't use it either, to keep the code consistenct. Here are suggestions for alternative names:

`bFirstUpdate` -> `isFirstUpdate`
`bTls` -> `tlsIsSupported`
`bHTTPS` -> `useHttps`
`bCheckCertificate` -> `mustCheckCertificate`
`bCheckHostName` -> `mustCheckHostName`
`bSafeTLSSupport` -> `safeTlsIsSupported`

ducasp replied:
Done :)

Konamiman commented on SRC/NETWORK/hget.c:
```diff
@@ -422,13 +444,25 @@ void InitializeTcpipUnapi()
         Terminate("This TCP/IP implementation does not support active TCP connections.");
     }
-    
+	if(regs.Bytes.D & 16) //fifth bit is set?
```
Please convert that 16 to a constant, e.g. `TCPIP_CAPAB_VERIFY_CERTIFICATE`. This way the code reads better and you don't need the comment.

ducasp replied:
Done :)

Konamiman commented on SRC/NETWORK/hget.c:
```diff
@@ -123,32 +135,32 @@ enum TcpipErrorCodes {
 const char* strUsage=
-    "Usage: hget <file URL>|<local file with the URL>|con [/l:<local file name>]\r\n"
-    "            [/c] [/v] [/h] [/x:<headers file name>] [/a:<user>:<password>]\r\n"
+    "Usage: hgetf <file URL>|<local file with the URL>|con [/l:<local file name>]\r\n"
```
Why `hgetf`?

ducasp replied:
Forgot to remove it from my previous customized version :)

Konamiman commented on SRC/NETWORK/hget.c:
```diff
@@ -123,32 +135,32 @@ enum TcpipErrorCodes {
-    "HTTP file downloader 1.1\r\n"
-    "By Konamiman, 1/2011\r\n"
+    "HTTP file downloader 1.3\r\n"
+    "By Konamiman, 2011 v1.3 Oduvaldo (ducasp@gmail.com) 07/2019\r\n"
```
Alternative wording that imho will read better:

By Oduvaldo (ducasp@gmail.com), 7/2019
Based oh HGET 1.1 by Konamiman

ducasp replied:
I've done it, but on a single line, otherwise when printing instructions it would scroll down and not show the application name/version.

Konamiman commented on SRC/NETWORK/hget.c:
```diff
@@ -1409,9 +1487,49 @@ void CreateLocalFile()
+			{
+				bFirstUpdate=false;
+				print("\r[                         ]");
+				__asm
```
Please move this chunk of assembler to its own function with a descriptive name, this way the code will be easier to read.

ducasp replied:
Done :)

ducasp commented:
I've added TFTP as well, this version will not wait extra tick after TCPIP_WAIT and changes a bit the way to print transfer progress (do not print the .0 or .5, just print the text string first with the unit before the numbers, and then on each iteration just print the numbers, going back before doing so, this is faster and allow faster transfers). Also changed the routine to convert size to ASCII to one that is simpler and faster, that one was overkill and not as fast I believe as it had to check lots of possibilities and use a lot IX  registers, when no other options are being used... :-)

Konamiman commented:
I think you missed my email 😅

Everything looks good regarding HGET, but I'd prefer SNTP and TFTP changes to get each its own pull request. And their own code branches too, rather than everything coming from your master branch, please.

ducasp commented:
I believe I did it before your email 😅 no problems, I will do it later
today or tomorrow, I'm not familiar with git at all, I've used TFS and
ClearCase and other tools before, and I was not using command line got, I
was using the web interface. 😬

Thanks for the patience and that's a great learning for me, as soon as I
get back to the computer to look at MSX related stuff I will make as you
described.

Best regards,
Oduvaldo


Em sáb, 20 de jul de 2019 02:29, Néstor Soriano <notifications@github.com>
escreveu:

> I think you missed my email 😅
>
> Everything looks good regarding HGET, but I'd prefer SNTP and TFTP changes
> to get each its own pull request. And their own code branches too, rather
> than everything coming from your master branch, please.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Konamiman/MSX/pull/7?email_source=notifications&email_token=ADEEE2FHCEOQIUFTTXONLXTQAKPC7A5CNFSM4IFB44S2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2NHBTQ#issuecomment-513437902>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADEEE2G446FN66JR3UCEBPDQAKPC7ANCNFSM4IFB44SQ>
> .
>

Konamiman commented:
Well since you added new commits, the instructions I gave you to delete the last one won't work anymore. I think an interactive rebase would be the way to go to remove the extra commits, but if you aren't familiar with Git it's not goint to be an easy task. I can try to do it myself if you allow me to edit your pull request: https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork

ducasp commented:
I've let this permission  to edit the request and files on when I have started the pull request. So you should be able to edit it. 🙂

Konamiman commented:
I screwed up something and ended up deleting all your commits - although I still have them locally, as well as backup copies of all the files. I'll try to fix it.

Konamiman commented:
Could you please add me as a collaborator to your repository?

ducasp commented:
I've done that, and don't worry, if you think it is better perhaps we can
remove this pull request and I can start doing it as you recommended. 🙂

Em sáb, 20 de jul de 2019 10:09, Néstor Soriano <notifications@github.com>
escreveu:

> Could you please add me as a collaborator to your repository?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Konamiman/MSX/pull/7?email_source=notifications&email_token=ADEEE2FKGLTSFJ2F254TCJTQAME65A5CNFSM4IFB44S2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2NN7HY#issuecomment-513466271>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADEEE2DTO2AWQZSQE4GDNYDQAME65ANCNFSM4IFB44SQ>
> .
>

Konamiman commented:
Ok, fixed. Now you can go and create separate branches and pull requests for your changes in TFTP an SNTP.

ducasp commented:
Not sure if I did it right, both the SNTP and tftp push request tried to
get hget.c as well... I'll try to look at it tomorrow...

Em dom, 21 de jul de 2019 09:16, Néstor Soriano <notifications@github.com>
escreveu:

> Ok, fixed. Now you can go and create separate branches and pull requests
> for your changes in TFTP an SNTP.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Konamiman/MSX/pull/7?email_source=notifications&email_token=ADEEE2EOWQLKFYBDMUHOGCTQARHTRA5CNFSM4IFB44S2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2OCCVA#issuecomment-513548628>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADEEE2FNBSIEYCFX5YG5CBDQARHTRANCNFSM4IFB44SQ>
> .
>

Konamiman commented:
Looks like you have committed the SNTP and TFTP changes to master, and after that you have created the new branches. Therefore now all of the commits are present in all of the branches.

I'll try to fix the mess. Meanwhile please sit down, relax and read this: https://learngitbranching.js.org 🙂

Konamiman commented:
Closing this, replaced with https://github.com/Konamiman/MSX/pull/10
